src/flint/flint_base/flint_base.pyx

```diff
@@ -10,78 +12,78 @@ cdef class flint_elem:
-#         if not coeffs:
-#             return "0"
-#         s = []
-#         coeffs = enumerate(coeffs)
-#         if not ascending:
-#             coeffs = reversed(list(coeffs))
-#         for i, c in coeffs:
-#             if c == "0":
-#                 continue
-#             else:
-#                 if c.startswith("-") or (" " in c):
-#                     c = "(" + c + ")"
-#                 if i == 0:
-#                     s.append("%s" % c)
-#                 elif i == 1:
-#                     if c == "1":
-#                         s.append("x")
-#                     else:
-#                         s.append("%s*x" % c)
-#                 else:
-#                     if c == "1":
-#                         s.append("x^%s" % i)
-#                     else:
-#                         s.append("%s*x^%s" % (c, i))
-#         return " + ".join(s)
-
-#     def roots(self, **kwargs):
-#         """
-#         Isolates the complex roots of *self*. See :meth:`.acb_poly.roots`
-#         for details.
-#         """
-#         # TODO: 
-#         # To avoid circular imports, we import within the method
-#         from XXX.XXX.acb_poly import acb_poly
-#         return acb_poly(self).roots(**kwargs)
+
+
+cdef class flint_poly(flint_elem):
+    """
+    Base class for polynomials.
+    """
+
+    def __iter__(self):
+        cdef long i, n
+        n = self.length()
+        for i in range(n):
+            yield self[i]
+
+    def coeffs(self):
+        return list(self)
+
+    def str(self, bint ascending=False):
+        """
+        Convert to a human-readable string (generic implementation for
+        all polynomial types).
+
+        If *ascending* is *True*, the monomials are output from low degree to
+        high, otherwise from high to low.
+        """
+        coeffs = [str(c) for c in self]
+        if not coeffs:
+            return "0"
+        s = []
+        coeffs = enumerate(coeffs)
+        if not ascending:
+            coeffs = reversed(list(coeffs))
+        for i, c in coeffs:
+            if c == "0":
+                continue
+            else:
+                if c.startswith("-") or (" " in c):
+                    c = "(" + c + ")"
+                if i == 0:
+                    s.append("%s" % c)
+                elif i == 1:
+                    if c == "1":
+                        s.append("x")
+                    else:
+                        s.append("%s*x" % c)
+                else:
+                    if c == "1":
+                        s.append("x^%s" % i)
+                    else:
+                        s.append("%s*x^%s" % (c, i))
+        return " + ".join(s)
+
+    def roots(self):
+        """
+        Depreciated function.
+        
+        To recover roots of a polynomial, first convert to acb:
+
+        acb_poly(input_poly).roots()
+        """
+        warn('This method is deprecated. Please instead use acb_poly(input_poly).roots()', DeprecationWarning)
+        
+
```

Comment:
Warnings are useful only if we want to preserve existing behaviour while giving out the warning. This does not return the roots that would have previously been returned so existing behaviour is not preserved.

I think it is better just to raise an error here rather than give out a warning and implicity return None.

Reply:
Also if giving an error (or not preserving existing behaviour) then it does not really make sense to refer to this as "deprecated": rather it is "unsupported" or "disallowed" or something like that.

Reply:
Yes, I agree, I'll swap this for
```py
NotImplementedError('This method has been deprecated. Please instead use acb_poly(input_poly).roots()')
```

Reply:
Also we should add a CHANGELOG section at the bottom of the README with something like:
```
CHANGELOG
---------

0.5.0

- gh-63: The `roots` method of `fmpz_poly`, `fmpq_poly` and `nmod_poly` (others?) is no longer supported. Use `acb_roots(p).roots()` to get the old behaviour of returning the roots as `acb`.

```

Reply:
Maybe the simplest thing to do is simply say it's not supported and point to the `acb_poly()` then...

Reply:
`fmpz_poly` still has roots, but they're complex roots, which is weird. `fmpq_poly` asks for roots of the numerator, which I believe is of type `fmpz_poly`, so the only ones currently broken I think are `arb_poly` and `nmod_poly`

```
>>> acb_poly([1,2]).roots()
[-0.500000000000000]
>>> arb_poly([1,2]).roots()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "src/flint/flint_base/flint_base.pyx", line 75, in flint.flint_base.flint_base.flint_poly.roots
    raise NotImplementedError('This method is no longer supported. To recover the complex roots first convert to acb_poly')
NotImplementedError: This method is no longer supported. To recover the complex roots first convert to acb_poly
>>> fmpz_poly([1,2]).roots()
[(-0.500000000000000, 1)]
>>> fmpq_poly([1,2]).roots()
[(-0.500000000000000, 1)]
>>> nmod_poly([1,2], 11).roots()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "src/flint/flint_base/flint_base.pyx", line 75, in flint.flint_base.flint_base.flint_poly.roots
    raise NotImplementedError('This method is no longer supported. To recover the complex roots first convert to acb_poly')
NotImplementedError: This method is no longer supported. To recover the complex roots first convert to acb_poly
```